rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,176 @@
+separate resources managed by Teleport and those managed by the user manually.
+
+Resource labels will be taken into account when doing the reconciliation - users will be able to match different
+Kubernetes clusters for different KubeProvision resources. If a KubeProvision resource doesn't have any labels defined it
+will not match any clusters, effectively being disabled.
+
+In order to be able to reconcile the state, Teleport will need to perform CRUD operations on Kubernetes RBAC resources.
+When performing CRUD operations, Teleport will impersonate the "default-cluster-admin" group, that was described in the previous
+section. This means that only newly enrolled
+Kubernetes clusters, or cluster where the user performed a manual upgrade of permissions will be in scope of provisioning the resources,
+since otherwise the "default-cluster-admin" role won't be available.
+
+## Security
+
+The introduction of this functionality does not directly impact the security of Teleport
+itself, however it introduces a new vector of attack for a malicious actor.
+Teleport user with sufficient permissions to create/edit KubeProvision resources
+will be able to amend RBAC setup on all Kubernetes clusters enrolled in Teleport.
+We will emphasize in the user documentation the need to be vigilant when giving out
+the necessary permissions.
+
+Even though we will always run the reconciliation loop, by default it will be a no-op, since three will
+be no resources to provision, so users need to explicitly create KubeProvision resources to start actively using this new feature.
+We also will explicitly require labels to be present for the resource to be provisioned to the Kubernetes cluster, making it 
+more difficult to accidentally misuse the feature.
+
+## Alternative
+
+Alternatively, we could take a bit of a different approach regarding permissions and default roles exposure.
+We could directly add permissions required to perform CRUD operations on Kubernetes RBAC resources to 
+Teleport kube agent/service account credentials on enrollment. We would also ship Teleport with a default KubeProvision 
+resource that defines role binding for the default user facing roles for edit and view. This default resource will not have any
+labels defined, so it would not be provisioned anywhere by default. If users wanted to enable the exposure of default view/edit roles,
+they would need to set labels on that resource. They can then use these roles with the fine-grained RBAC definitions for Kubernetes
+in Teleport roles. This is a more conservative scenario that will require more explicit decision-making from the user. 
+To expose the default Kubernetes user-facing roles, users would need to add labels to the default resource we provide,
+and KubeProvisioning will only be active for clusters where Teleport has the required permissions added to its credentials.
```

Comment:
With this approach we would have a single way of provisioning kube RBAC, this reduces the overall setup complexity and I really like it.

If, instead of having users add labels on the KubeProvision resources we ship by default, we had default labels and control which KubeProvision is pulled via selections configured on the agent? We can have discover and guides set the selectors by default we could achieve the same "run 1 command, everything works" experience. This would also not be a surprise change for existing users.

For example if we ship the default KubeProvision RBAC labeled with `teleport.dev/default: {ro,rw}`, we can have the users deploy the agent with

```
roles: kube,app,discovery
authToken: test-auth-token
proxyAddr: teleport.example.com:443
kubeClusterName: test
kubeProvisionResources:
  - labels:
      "teleport.dev/default": "*" 
enterprise: true
```

The agent would automatically pick up out default kube RBAC and new users would be able to enjoy kueb access working out of the box. This would also be an additional security layer: clusters are pulling the kube provision resources they want. This approach could also reuse the same selector logic and syntax we have for apps/dbs. Users are already familiar with this.

Reply:
I'm not sure if it's an easier setup. For the default roles exposure I think overall it we should shift paradigm of Teleport usage where we think of it as a default (those "quick-access" groups are available by default), so it makes sense just create those bindings on on cluster enrolment, we want to make Day 1 experience for users simple and easy, as Sasha said last Friday on all hands. And kube provision is more of a Day 2 feature, for more experienced users. I think it's cleaner just to always create bindings. And we could have problems with users accidentally/by mistake deleting/changing default kube_provision resources and breaking all access.

For `kubeProvisionResources`, we could go this way, but it also seems a bit of an overcomplication, it's probably more flexible and easier to use if you target clusters and can make changes centrally with `tctl` instead of on every agent (if you need some change) But it does look a bit more secure 🤔